src/main/java/cz/smarteon/loxone/LoxoneIconRegistry.java

```diff
@@ -0,0 +1,68 @@
+package cz.smarteon.loxone;
+
+import cz.smarteon.loxone.message.ImageCommand;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+public class LoxoneIconRegistry implements CommandRequestResponseListener<ByteBuffer> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LoxoneIconRegistry.class);
+
+    private final Map<String, ByteBuffer> iconMap = new HashMap<>();
+    private LoxoneWebSocket loxoneWebSocket;
+
+    private CountDownLatch iconLatch;
+
+    @Override
+    public void registerWebSocket(final @NotNull LoxoneWebSocket loxoneWebSocket) {
+        this.loxoneWebSocket = loxoneWebSocket;
+    }
+
+    @Nullable
+    public synchronized ByteBuffer getIcon(final @NotNull String iconUuid) {
+        requireNonNull(loxoneWebSocket, "LoxoneWebSocket should be available!");
+        if (iconMap.containsKey(iconUuid)) {
+            return iconMap.get(iconUuid);
+        } else {
+            iconLatch = new CountDownLatch(1);
+            loxoneWebSocket.sendCommand(ImageCommand.genericControlCommand(iconUuid));
+            try {
+                final int timeout = loxoneWebSocket.getAuthTimeoutSeconds() * loxoneWebSocket.getRetries() + 1;
+                if (iconLatch.await(timeout, TimeUnit.SECONDS)) {
+                    return iconMap.get(iconUuid);
+                } else {
+                    LOG.error("Loxone icon wasn't fetched within timeout");
```

Comment:
the bad thing here is that if no file is received the command is pending in LoxoneWebsocket creating incosistency there and possibly breaking all the communication - this is not your issue, rather the design issue of the whole library.

Could you please find out what happens when somebody request a file which does not exist? Does miniserver answer anything or just nothing?

Reply:
I've tried this and found that the miniserver responds with following message:

`{"LL": { "control": "IconsFilled/unknown.svg", "value": "Bad request", "Code": "400"}}`

This is a message with TEXT format, so picked up by the `processMessage(final String message)`where it fails to convert the message to the ByteBuffer format specified by the command. I've tried to use LoxoneMessage<JsonValue> as response type, but that causes the LoxoneAuth command listener to process the command. This class thinks the current token is invalid and discards this.

Not entirely clear to me if the response type of the command shoud be LoxoneMessage<JsonValue>, this is only applicable if the command results in an error. When the file is found it answers with a FILE type command.

Reply:
thx for trying, I will look closer on that